tests/transaction_tests.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+
+import datetime
+import os
+import sys
+import struct
+import socket
+import tempfile
+import time
+
+import logging
+logging.basicConfig(level=logging.INFO, format="%(message)s")
+logger = logging.getLogger('program')
+
+import threading
+import subprocess
+
+import inspect
+
+from m3.ice import ICE
+import m3.ice_simulator
+import m3.m3_ice
+
+class TestICE(object):
+    class TestFailedException(Exception):
+        #logger.info('='*42 + '\nTEST FAILED\n' + '='*42)
+        pass
+
+    @classmethod
+    def setup_class(cls, ):
+       
+        #make sure we start with a fresh serial port
+        serial_port=m3.ice_simulator._FAKE_SERIAL_SIMULATOR_ENDPOINT
+        if os.path.exists(serial_port):
+            os.remove(serial_port)
+
+        cls.sim_thread = None
+        filedir = os.path.dirname(os.path.realpath(__file__))
+
+        os.environ['ICE_NOSLEEP'] = '1'
+        sim_args = m3.ice_simulator.Simulator.get_parser().parse_args([
+                                        #'-s ' + serial_port, 
+                                        '-t ' + filedir +'/transactions/timeout.trx'])
+
+        cls.sim_thread = threading.Thread(
+                target=m3.ice_simulator.Simulator(args=sim_args).run,
+                name='fake_ice',
+                )
+        cls.sim_thread.daemon = True
+        cls.sim_thread.start()
+
+        time.sleep(0.5)
+
+        #cls.ice = ICE()
+        #cls.ice.connect(com2)
+        #cls.ice.connect(serial_port)
+
+    @classmethod
+    def teardown_class(cls):
+        m3.ice_simulator.destroy_fake_serial()
+
+    def test_timeout(self):
+        logger.info("Testing Timeout Feature") 
+
+        serial_port=m3.ice_simulator._FAKE_SERIAL_CONNECTTO_ENDPOINT
+        print ('Using ' + str(serial_port))
+        self.driver = m3.m3_ice.m3_ice(['--debug',
+                                    '-s '+ serial_port,
+                                    'snoop'])
+        self.driver.cmd_snoop()
+        
+        
+        # I have no idea how this works.....
```

Comment:
Hahaha.. I'm no expert, but my understanding is that `nose` from Python does lots of magic based on the names of things. It'll scan every file in the `tests/` folder, for anything with the string `test` in the name it'll treat that as a unit testing file. Any class in that file with `test` in the name it'll assume is written for it to run. The magic functions `setup_class` and `teardown_class` are things that nose will call for you before running any of your tests. Any function with the string `test` in it nose will automatically run.

It's all magic to me and all matching of magic well-known strings IIRC